File: core/src/main/java/com/emarsys/core/util/JsonUtils.java
package com.emarsys.core.util;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class JsonUtils {

    public static JSONObject fromMap(Map<String, Object> map) {
        Assert.notNull(map, "Map must not be null!");

        JSONObject result = new JSONObject();

        try {
            for (Map.Entry<String, Object> entry : map.entrySet()) {
                String key = entry.getKey();
                Object value = entry.getValue();

                if (value instanceof Map) {
                    result.put(key, JsonUtils.fromMap((Map<String, Object>) value));
                } else if (value instanceof List) {
                    result.put(key, JsonUtils.fromList((List<Object>) value));
                } else {
                    result.put(key, value);
                }
            }
        } catch (JSONException je) {
            throw new IllegalArgumentException(je);
        }

        return result;
    }

    public static JSONArray fromList(List<Object> list) {
        Assert.notNull(list, "List must not be null!");

        JSONArray result = new JSONArray();

        for (Object item : list) {
            if (item instanceof List) {
                result.put(JsonUtils.fromList((List<Object>) item));
            } else if (item instanceof Map) {
                result.put(JsonUtils.fromMap((Map<String, Object>) item));
            } else {
                result.put(item);
            }
        }

        return result;
    }

    public static JSONObject merge(JSONObject... jsonObjects) {
        validateArgument(jsonObjects);

        JSONObject result = new JSONObject();

        for (JSONObject jsonObject : jsonObjects) {
            if (jsonObject != null) {
                Iterator<String> iterator = jsonObject.keys();
                while (iterator.hasNext()) {
                    String key = iterator.next();
                    try {
                        result.put(key, jsonObject.get(key));
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }
            }
        }
        return result;
    }

    public static Map<String, String> toFlatMap(JSONObject jsonObject) {
        Assert.notNull(jsonObject, "JsonObject must not be null!");

        Map<String, String> result = new HashMap<>();
        Iterator<String> iterator = jsonObject.keys();
        while (iterator.hasNext()) {
            try {
                String key = iterator.next();
                String value = jsonObject.getString(key);
                result.put(key, value);
            } catch (JSONException ignore) {
            }
        }
        return result;
    }

    private static void validateArgument(JSONObject[] jsonObjects) {
        if (jsonObjects.length == 0) {
            throw new IllegalArgumentException("Argument must not be empty array!");
        }

        int nullCount = 0;
        for (JSONObject jsonObject : jsonObjects) {
            if (jsonObject == null) {
                nullCount++;
            }
        }
        if (nullCount == jsonObjects.length) {
            throw new IllegalArgumentException("Argument must contain at least one not null element!");
        }
    }

}

File: core/src/main/java/com/emarsys/core/util/JsonUtils.kt
package com.emarsys.core.util

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

object JsonUtils {

    fun JSONArray.toMutableList(): MutableList<JSONObject> = MutableList(length(), this::getJSONObject)

    @JvmStatic
    fun fromMap(map: Map<String, Any>): JSONObject {
        val result = JSONObject()
        try {
            for ((key, value) in map) {
                when (value) {
                    is Map<*, *> -> {
                        result.put(key, fromMap(value as Map<String, Any>))
                    }
                    is List<*> -> {
                        result.put(key, fromList(value as List<Any?>))
                    }
                    else -> {
                        result.put(key, value)
                    }
                }
            }
        } catch (je: JSONException) {
            throw IllegalArgumentException(je)
        }
        return result
    }

    @JvmStatic
    fun fromList(list: List<Any?>): JSONArray {
        val result = JSONArray()
        for (item in list) {
            when (item) {
                is List<Any?> -> {
                    result.put(fromList(item))
                }
                is Map<*, *> -> {
                    result.put(fromMap(item as Map<String, Any>))
                }
                else -> {
                    result.put(item)
                }
            }
        }
        return result
    }

    @JvmStatic
    fun merge(vararg jsonObjects: JSONObject?): JSONObject {
        validateArgument(arrayOf(*jsonObjects))

        val result = JSONObject()
        for (jsonObject in jsonObjects) {
            jsonObject?.keys()?.asSequence()?.forEach {
                try {
                    result.put(it, jsonObject[it])
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }
        }
        return result
    }

    @JvmStatic
    fun toFlatMap(jsonObject: JSONObject): Map<String, String> {
        return jsonObject.keys().asSequence().associateBy({ key -> key }) { key ->
            try {
                jsonObject.getString(key).convertNullStringValueToNull()
            } catch (ignore: JSONException) {
                null
            }
        }.mapNotNull { it.value?.let { value -> it.key to value } }.toMap()
    }

    @JvmStatic
    fun toMap(jsonObject: JSONObject): Map<String, Any> {
        val result = mutableMapOf<String, Any>()
        jsonObject.keys().asSequence().forEach {
            when (val value = jsonObject.get(it)) {
                is JSONObject -> {
                    result[it] = toMap(value)
                }
                is JSONArray -> {
                    result[it] = toList(value)
                }
                else -> {
                    result[it] = value
                }
            }
        }
        return result
    }

    @JvmStatic
    fun toList(jsonArray: JSONArray): List<Any> {
        val result = mutableListOf<Any>()
        for (i in 0 until jsonArray.length()) {
            when (val value = jsonArray.get(i)) {
                is JSONObject -> {
                    result.add(toMap(value))
                }
                is JSONArray -> {
                    result.add(toList(value))
                }
                else -> {
                    result.add(value)
                }
            }
        }
        return result
    }

    @JvmStatic
    fun toFlatMapIncludingNulls(jsonObject: JSONObject): Map<String, String?> {
        return jsonObject.keys().asSequence().associateBy({ key -> key }) { key ->
            try {
                jsonObject.getString(key).convertNullStringValueToNull()
            } catch (ignore: JSONException) {
                null
            }
        }
    }

    private fun validateArgument(jsonObjects: Array<JSONObject?>) {
        require(jsonObjects.isNotEmpty()) { "Argument must not be empty array!" }
        var nullCount = 0
        for (jsonObject in jsonObjects) {
            if (jsonObject == null) {
                nullCount++
            }
        }
        require(nullCount != jsonObjects.size) { "Argument must contain at least one not null element!" }
    }
}

fun JSONObject.getNullableString(key: String): String? {
    return if (this.isNull(key)) null else this.getString(key)
}

fun JSONObject.getNullableLong(key: String): Long? {
    return if (this.isNull(key)) null else this.getLong(key)
}

private fun String.convertNullStringValueToNull(): String? {
    return if (this == "null") null else this
}
